### ISSUE TITLE
File locked with read and delete permissions behave differently between client and server OS

### ISSUE THREAD

**Issue by AdmiringWorm** [chocolatey/choco#3144]
### Checklist

- [X] I have verified this is the correct repository for opening this issue.
- [X] I have verified no other issues exist related to my problem.
- [X] I have verified this is not an issue for a specific package.
- [X] I have verified this issue is not security related.

### What You Are Seeing?

On a client system with the latest code we are working against allows a backup folder to be deleted if it contains a file that has been read and delete locked, even if that file was moved from a different location.
On a server it is not possible to delete this file when it is read and delete locked, it is still allowed to be moved to any new location.

### What is Expected?

The expectation is that behavior is consistent, and it will be removed on both client and server or kept on both.

### How Did You Get This To Happen?

1. Install the package `upgradepackage` version `1.0.0`.
2. Lock a file in the directory using `$LockedFile = [System.IO.File]::Open("C:\ProgramData\chocolatey\lib\upgradepackage\tools\chocolateyInstall.ps1", "OpenOrCreate", "ReadWrite", "Read,Delete")`.
3. Upgrade the package `upgradepackage` (ignore any warnings).
4. Verify whether the directory `lib-bkp\upgradepackage\1.0.0` exist or not (will not exist on client systems, but exist on server systems).

### System Details

* Operating System: Windows 10 and Windows 2019
* Windows PowerShell version: 5.1.19041.2673 (Server: 5.1.17763.1852)
* Chocolatey CLI Version: 2.0.0-vNext
* Chocolatey Licensed Extension version: N/A
* Chocolatey License type: N/A
* Terminal/Emulator: PowerShell

### Installed Packages

```bash
chocolatey 2.0.0-PROJ615bef
installpackage 1.0.0
upgradepackage 1.1.0
```

### Output Log

```bash
https://gist.github.com/AdmiringWorm/cf023056f77e45be24d19a196b4c686d
```

### Additional Context

The included log includes the full log from upgrading 1.2.1 until the end of the v2 testing.

┆Issue is synchronized with this [Clickup task](https://app.clickup.com/t/868m0agke) by [Unito](https://www.unito.io)

**Comment by corbob** [comment]
I don't think this is a difference between Server and Client OS, and more between 2019 and 2022+ code bases... And at the very least in my minor testing: if you do the locking mentioned in the description, on Server 2019 you can't remove the file with `Remove-Item`, but on 2022 you can.

**Comment by corbob** [comment]
### Root cause / behaviour summary

This is a Windows file-deletion behaviour difference, not a Chocolatey bug. On a forced reinstall, Chocolatey backs the existing package up to `lib-bkp` while a package file is held open with a *delete-granting* share (`FileShare.Delete` / `Read, Delete`), then cleans `lib-bkp` up. Whether that cleanup succeeds depends on the OS:

- **Windows Server 2019 and earlier retain** `lib-bkp` — the open, delete-shared file blocks removal of its directory.
- **Windows Server 2022 and newer, and client Windows 10, remove** `lib-bkp`.

Verified in Test Kitchen (build 735) across Server 2016/2019/2022/2025, and consistent with the original report of a Windows 10 client deleting the backup.

By OS build, the last release we observe retaining the backup is Windows Server 2019 / Windows 10 1809 (build `17763`); the behaviour is present from the next release, Windows 10 1903 (build `10.0.18362`), onward — and on Server 2022 (`20348`) and later. Build numbers: [Windows 10 release info](https://learn.microsoft.com/en-us/windows/release-health/release-information), [Windows Server release info](https://learn.microsoft.com/en-us/windows-server/get-started/windows-server-release-info).

The likely cause is a change in how Windows removes a file opened with `FILE_SHARE_DELETE` — widely understood to be POSIX-style deletes becoming the default in newer Windows — discussed in [golang/go#78072](https://github.com/golang/go/issues/78072) and [dotnet/runtime#32737](https://github.com/dotnet/runtime/issues/32737). Note that Microsoft documents the *opt-in* flag ([`FILE_DISPOSITION_INFORMATION_EX`](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddk/ns-ntddk-_file_disposition_information_ex)) but I could not find an authoritative Microsoft statement that the *default* changed at a specific build — so this is treated as observed behaviour, not a documented change.
